libyul/optimiser/MemoryEscalator.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/MemoryEscalator.h>
+#include <libyul/optimiser/CallGraphGenerator.h>
+#include <libyul/optimiser/FunctionCallFinder.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/StackToMemoryMover.h>
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/AsmData.h>
+#include <libyul/CompilabilityChecker.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+#include <libyul/Object.h>
+#include <libyul/Utilities.h>
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/CommonData.h>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+MemoryOffsetAllocator::MemoryOffsetAllocator(
+	map<YulString, FunctionStackErrorInfo> const& _functionStackErrorInfo,
+	set<YulString> const& _functionsInCycle,
+	map<YulString, set<YulString>> const& _callGraph
+): m_functionStackErrorInfo(_functionStackErrorInfo), m_functionsInCycle(_functionsInCycle), m_callGraph(_callGraph)
+{
+}
+
+bool MemoryOffsetAllocator::run() { return run({}); }
+uint64_t MemoryOffsetAllocator::requiredSlots() const {	return m_nextAvailableSlot.at({}); }
+
+bool MemoryOffsetAllocator::run(YulString v)
+{
+	yulAssert(!m_nextAvailableSlot.count(v), "");
+	uint64_t& n = m_nextAvailableSlot[v];
+	if (m_callGraph.count(v))
```

Comment:
The call graph is complete, so you do not need this condition.

Reply:
It's not! The fact that it isn't annoyed me quite a few times...

Reply:
It contains calls to builtins, which doesn't hurt the algorithm, but the builtins themselves have no entries in the callgraph again - I could of course filter out builtins, but compared to that this seems simpler to me...